MVC UI migration to AdminLTE 3

iyilm4z opened the pull request:
Fixes issues [4838](https://github.com/aspnetboilerplate/aspnetboilerplate/issues/4838) and [4943](https://github.com/aspnetboilerplate/aspnetboilerplate/issues/4943) in ABP's framework repo.

Extra improvements:

- Recursive SideBarMenu
- [Datatables.net](https://github.com/DataTables/DataTables) integration
- [libman](https://github.com/aspnet/LibraryManager) integration
- Advanced filters infrastructure for listing pages
- Pagination feature for listing pages
- Added asterisk to required fields


Possible Road map:
- Basic Dashboard with [Chart.js](https://github.com/chartjs/Chart.js)
- Theme Customization(horizontal menu, changing colors of layout components, etc.)

totpero commented on aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Account/Login.cshtml:
```diff
@@ -23,45 +13,52 @@
-        </form>
+        </div>
+        <div class="col-4">
+            <button type="submit" id="LoginButton" class="btn btn-primary btn-block">Sign In</button>
```
Use localizatin @L("LogIn")

iyilm4z replied:
done. thanks!

totpero commented on aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Account/Login.cshtml:
```diff
@@ -23,45 +13,52 @@
+            <div class="icheck-primary">
+                <input type="checkbox" id="RememberMe" name="RememberMe" value="true">
+                <label for="RememberMe">
+                    Remember Me
```
Use localizatin @L("RememberMe")

iyilm4z replied:
done. thanks!

iyilm4z commented on aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Account/_Layout.cshtml:
```diff
@@ -5,121 +5,99 @@
+    <link href="~/css/style.css" rel="stylesheet" />
 
-    <environment names="Staging,Production">
-        <link rel="stylesheet" href="~/view-resources/Views/_Bundles/account-layout.min.css" asp-append-version="true" />
```
Thanks for your review. I'll consider on this.

cathalbahn replied:
I have made this change along with a few others. See: https://github.com/iyilm4z/module-zero-core-template/pull/1

iyilm4z replied:
@cathalbahn dude, thanks for your effort but i can't merge your this pr, bcoz it's contains many different changes beside environment tags.

cathalbahn replied:
@iyilm4z 
The "Environment Tags" depends on the "Bundle Config" and "Compiler Config".

That leaves the following changes:

- Added Alert Manager
- Replaced Right User Area with Component
- Added Dashboard with Chart
- Removed Unused RightSideBar
- Removed Unused SideBarNav
- Html Fixes

Do you want them all removed from the PR?

iyilm4z replied:
@cathalbahn yes please. Let me see changes one by one. It makes it easier for me to review.

iyilm4z replied:
@cathalbahn I merged your pr. Thanks for it.

demirmusa commented:
Hi @iyilm4z. 
First of all, I just want to thank you for that big pull request.
I am going to check them all out and write you where to fix if it needs any fixes.

iyilm4z commented:
@demirmusa thanks, i'm waiting your feedbacks.

iyilm4z commented on aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Controllers/HomeController.cs:
```diff
@@ -1,15 +1,32 @@
+        private readonly IUserAppService _userAppService;
+        private readonly IRoleAppService _roleAppService;
+
+        public HomeController(ITenantAppService tenantAppService,
```
Yes you are right, i would use these services for a basic dashboard with chart.js but i changed my mind and wanted to do it later on. i fixed it with a new change.

iyilm4z commented on aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Controllers/RolesController.cs:
```diff
@@ -22,23 +21,21 @@ public RolesController(IRoleAppService roleAppService)
             var permissions = (await _roleAppService.GetAllPermissions()).Items;
             var model = new RoleListViewModel
             {
-                Roles = roles,
```
Roles property was no needed anymore because i used js service proxies with Datatables.net to bind roles. [Here](https://github.com/iyilm4z/module-zero-core-template/blob/352a70cea875437b17ea66498df809f2dbbacd3a/aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/wwwroot/view-resources/Views/Roles/Index.js#L17) you can see it.

demirmusa replied:
Since you change lib folder name to libs, there are tons of js,css changes which I can not check. 

What about we revert folder name to lib, than change it's name in another pr after we more that one?

iyilm4z replied:
Sure dear friend, btw thanks for your informative reviewing. I'll do it with a new change set soon.

iyilm4z replied:
Done. [Here](https://github.com/aspnetboilerplate/module-zero-core-template/pull/489/commits/9450ce6c92131a040a55eb33f8aa5ea6fc6e0d7e) it is.

demirmusa replied:
I thought it would reduce the number of file changes, but there was hardly any number change since all css js files changed.
Sory about that 😄

iyilm4z replied:
:)

iyilm4z replied:
I reverted it. [Here](https://github.com/aspnetboilerplate/module-zero-core-template/pull/489/commits/07da58a78aab9888330f743a7194f9c400784c5f) it is.

iyilm4z commented on aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Controllers/TenantsController.cs:
```diff
@@ -19,16 +18,12 @@ public TenantsController(ITenantAppService tenantAppService)
 
-        public async Task<ActionResult> Index()
-        {
-            var output = await _tenantAppService.GetAllAsync(new PagedTenantResultRequestDto { MaxResultCount = int.MaxValue }); // Paging not implemented yet
```
The same reason, i used role service js proxy for this action as well.

iyilm4z replied:
Duplicate of answer 2

iyilm4z commented on aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Controllers/UsersController.cs:
```diff
@@ -22,17 +21,15 @@ public UsersController(IUserAppService userAppService)
 
         public async Task<ActionResult> Index()
         {
-            var users = (await _userAppService.GetAllAsync(new PagedUserResultRequestDto {MaxResultCount = int.MaxValue})).Items; // Paging not implemented yet
```
Duplicate of answer 2

iyilm4z commented on aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/libman.json:
```diff
@@ -0,0 +1,168 @@
+{
```
Because of [this](https://snyk.io/blog/bower-is-dead/) and [this](https://docs.microsoft.com/tr-tr/aspnet/core/client-side/libman/libman-vs?view=aspnetcore-3.1). Maybe i shouldn't do it? imho in following days we will be seeing libman so often.

demirmusa replied:
Yes, we may change it, but we usually use yarn. 
@ismcagdas What do you think?

ismcagdas replied:
I'm not sure about LibMan. Our experiences with Bundler & Minifier was not good and as I can see same team is behind the LibMan :). I will check LibMan and than we can decide.

iyilm4z commented on aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Shared/Layout/_Header.cshtml:
```diff
@@ -0,0 +1,5 @@
+﻿<nav class="main-header navbar navbar-expand navbar-white navbar-light">
+    @await Html.PartialAsync("_Header.LeftNavbar.cshtml")
+    @*@await Html.PartialAsync("_Header.SearchForm.cshtml")*@
```
No i didn't forget that. There are also other commented code blocks too. [Here](https://github.com/iyilm4z/module-zero-core-template/blob/352a70cea875437b17ea66498df809f2dbbacd3a/aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Shared/Layout/_Header.RightNavbar.cshtml#L5) are similar ones. I commented them out because users may want to implement them just by uncommenting them easily. I also would implement them later with a simple dashboard included. There would be a flag which makes theme on basic or advanced mode.

ismcagdas replied:
It would be better to remove not used code blocks as much as possible. I mentioned before but probably you forgot. It would be better if we can first port the existing template to Admin LTE and than make such enhancements one by one in the future. Otherwise, it gets harder to review such PRs for us.

iyilm4z replied:
So all i need to do is delete unneeded codes which my most favorite thing:) i'll do it with a new change set soon.

demirmusa replied:
Before you delete them you better store them in another branch if you think you will create pr later.

iyilm4z replied:
Sure. I already did it. Thanks for your attention.

iyilm4z replied:
@ismcagdas Previous theme had RightSideBar compnent to change theme color, in this new theme implementation we don't have that feature yet. However it's on my list of things to to soon. That's why there is no reason to have that component and it's app service&dtos, should i delete them too? Also I would name that new component which does this feature as ControlSidebar, because AdminLTE named it as it is.

iyilm4z commented on aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Account/_Layout.cshtml:
```diff
@@ -5,121 +5,99 @@
-    <environment names="Staging,Production">
-        <script src="~/view-resources/Views/_Bundles/account-layout.min.js" asp-append-version="true"></script>
-    </environment>
+        <script src="~/libs/jquery-validate/localization/messages_@Html.Raw(CultureInfo.CurrentUICulture.Name.Replace(" -", "_" ) + ".js" )" asp-append-version="true"></script>
```
Sorry It's my mistake, strange but it was working in this way too:) I fixed it with a new change.

iyilm4z commented on aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Shared/Layout/_Layout.cshtml:
```diff
@@ -0,0 +1,97 @@
+    <script src="~/libs/jquery-validate/jquery.validate.min.js"></script>
+    @if (CultureInfo.CurrentUICulture.Name != "en")
+    {
+        <script src="~/libs/jquery-validate/localization/messages_@Html.Raw(CultureInfo.CurrentUICulture.Name.Replace(" -", "_" ) + ".js" )" asp-append-version="true"></script>
```
the same with previous comment.

ismcagdas commented:
@iyilm4z thank you for your great contribution. I have reviewed the app UI and found non-functional parts. There is also one broken link. If you don't have time, I can fix those problems since they are minor things and easy to change. Just let me know.

iyilm4z commented:
@ismcagdas Thanks for your review, i'm appreciated. Please do the rest of the task as you want, because i'm extremely busy with my job(we have some serious problems to fix) and these busy working period will be continued until end of March. I also had some other cool feature ideas that i would like to do but unfortunately i can't. Again thanks for having time for me.

ismcagdas commented:
@iyilm4z thank you for your great contribution again 😄 , we will handle the minor changes.

ismcagdas commented:
@iyilm4z as I can see you are still working on this PR. To avoid conflicts, please let us know when you are done and we will make some changes.

iyilm4z commented:
@ismcagdas Today i had some leisure time and then i wanted to my final touches on this issue. 

**Extra improvements:**
1. I removed broken links, unused libs, commented codes and fixed broken links

2. I edited libman.json file for making lib names matching with names in package.json, with this way it will be easy to update new packages just by checking the libs in package.json of upsource.

3. I also formatted libman.json file for grouping abp libs and admin-lte libs, it's now cleaner to see dependencies of both, core and custom ones.

4. Alert manager ui moved to new partial view and made it better looking

5. Finally, updated to the [latest](https://github.com/ColorlibHQ/AdminLTE/releases) admin-lte version which is v3.0.2 and released 5 days ago

**Missing improvements:**

1. Updating abp libs to the latest

**Notes:**

1. Previous theme had theme color changing feature, this theme also have it, it even have better customization but i haven't done it yet:( But i will it when i have enough time later. This feature also had backend realted codes here in [this folder](https://github.com/aspnetboilerplate/module-zero-core-template/tree/master/aspnet-core/src/AbpCompanyName.AbpProjectName.Application/Configuration), so should we remove it or keep it staying? 

That was it from me so far:)

iyilm4z commented:
@ismcagdas most probably i did my final change named "jquery.validate localization file mismatch fixed", even i realize anything needed to do, i won't do it until you guys require it from me. This means now you can finally work on doing your own changes.

iyilm4z commented:
@demirmusa Any special reason why you replaced flag-icon with famfamfam-flags? flag-icon has better support, you can see [here](https://github.com/lipis/flag-icon-css).

ismcagdas commented:
@iyilm4z we just want to change the theme to AdminLTE. Other changes like migrating to flag-icon can be done in another PR. Keeping the PR as small as possible is better for us.

demirmusa commented:
@iyilm4z, our angular and react project also gets language icons from same database. We should also consider them. We can change them all in a split pr as @ismcagdas said.

iyilm4z commented:
@demirmusa @ismcagdas roger that:)

iyilm4z commented:
@demirmusa there is no file named **_Layout.js** it referenced as dev resource which courses error.
May i remove it? 

![bug](https://user-images.githubusercontent.com/9105939/73240928-8b482780-41b1-11ea-9c83-348d10623c02.JPG)

Also i realized i did a naming convention mistake with **_Layout.Alerts.cshtml** file, it should be named as **_Alerts.cshtml** and there are few enhancements on it i would like to do on it.

iyilm4z commented:
@demirmusa also what do you think about using partial views for styles and script references to simplify layout? Please look at [my repo](https://github.com/iyilm4z/admin-lte-dotnet) to see how. However @ismcagdas  probably think we should do it with another pr after this pr merged:)

demirmusa commented:
> @demirmusa there is no file named **_Layout.js** but it referenced as dev resource which courses error.
> May i remove it?

Yes it has no reference, forgotten from previous layout. You can remove it.

>@demirmusa also what do you think about using partial views for styles and script references to simplify layout?

I think we may leave it as it is. It would be better to do smaller jobs in smaller prs.
_________
I check most of the pr and I am about to finish my complete review then I will approve changes. _(Sorry for the glitch, we are busy these days)_.

demirmusa commented:
By the way @iyilm4z, you can create an issue(s) for your requested changes.

demirmusa commented:
@iyilm4z, Thank you for contributing. Great job.

@ismcagdas I approved changes but dont forget to consider about libman.

iyilm4z commented:
@ismcagdas @demirmusa i would like to start migrating angular client to admin lte when i have time. [Here](https://github.com/aspnetboilerplate/aspnetboilerplate/issues/4982) is related issue. However before i start working on it, i think this pr needs to be merged.

ismcagdas commented:
@iyilm4z you are right. I couldn't have time to finish my review. I will merge it at the weekend or monday.

iyilm4z commented:
@ismcagdas good news:) i'm waiting.

iyilm4z commented:
@demirmusa i realized that we should also implement [this](https://github.com/iyilm4z/module-zero-core-template/blob/4a893f0c478f8a985dbb6a2f6fd5f9da27f7bc0e/aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Shared/Layout/_Layout.cshtml#L99) missing logic for account layout [here ](https://github.com/iyilm4z/module-zero-core-template/blob/4a893f0c478f8a985dbb6a2f6fd5f9da27f7bc0e/aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Account/_Layout.cshtml#L103)

demirmusa commented:
> @demirmusa i realized that we should also implement [this](https://github.com/iyilm4z/module-zero-core-template/blob/4a893f0c478f8a985dbb6a2f6fd5f9da27f7bc0e/aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Shared/Layout/_Layout.cshtml#L99) missing logic for account layout [here ](https://github.com/iyilm4z/module-zero-core-template/blob/4a893f0c478f8a985dbb6a2f6fd5f9da27f7bc0e/aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Account/_Layout.cshtml#L103)

It was not implemented before that pr. 
https://github.com/aspnetboilerplate/module-zero-core-template/blob/master/aspnet-core/src/AbpCompanyName.AbpProjectName.Web.Mvc/Views/Account/_Layout.cshtml#L101-L104
We can do that in another pr. Can you please create another issue about that.

iyilm4z commented:
@demirmusa screenshots in readme still belong to admin bsb? we should also consider when to update them. Probably after angular client migration?

demirmusa commented:
> @demirmusa screenshots in readme still belong to admin bsb? we should also consider when to update them. Probably after angular client migration?

Thanks for the reminder. We can handle it after pr is merged.

ismcagdas commented:
@demirmusa could you change dropdown buttons on list pages (User, Role and Tenant) regular buttons. It should be like **[Edit] [Delete]**.

demirmusa commented:
Done 👍 
,
![image](https://user-images.githubusercontent.com/48536631/74531281-4821de80-4f3d-11ea-924f-9997970b3672.png)

iyilm4z commented:
@ismcagdas thanks for having time!
@demirmusa thanks for your collaboration!
